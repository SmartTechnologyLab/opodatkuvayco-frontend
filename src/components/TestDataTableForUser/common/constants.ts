import { Currency } from '@/constants/currencies'

export const notEditableColumns: string[] = [
  'total',
  'percent',
  'sale.uah',
  'sale.sum',
  'purchase.uah',
  'sale.uah',
  'purchase.sum',
  'sale.rate',
  'purchase.rate'
]

export enum TableSize {
  LG = 'Full Statement',
  SM = 'Short Statement'
}

export const currencyOptions = Object.values(Currency).filter((currency) => currency !== Currency.UAH)

export const tableSizeOptions = Object.values(TableSize)
