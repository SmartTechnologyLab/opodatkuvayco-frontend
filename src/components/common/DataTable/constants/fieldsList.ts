export const currencyFields: string[] = [
  'purchase.price',
  'purchase.commission',
  'sale.price',
  'sale.uah',
  'sale.commission'
]

export const calendarFields: string[] = ['purchase.date', 'sale.date']

export const numberFields: string[] = ['quantity', 'total']

export const textFields: string[] = ['ticker']
