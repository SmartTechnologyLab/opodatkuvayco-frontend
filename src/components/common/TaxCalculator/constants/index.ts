export const validChars = '0123456789+-/*.'

export const operations: string[] = ['*', '/', '+', '-']

export const values: Array<number | string> = [1, 2, 3, 4, 5, 6, 7, 8, 9, '+/-', 0, '.']
