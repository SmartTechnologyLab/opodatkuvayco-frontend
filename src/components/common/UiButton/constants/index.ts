export enum Icons {
  CHECK = 'pi pi-check',
  USER = 'pi pi-user',
  CROSS = 'pi pi-times',
  DELETE = 'pi pi-trash',
  PLUS = 'pi pi-plus'
}

export enum Severities {
  SECONDARY = 'secondary',
  HELP = 'help',
  DANGER = 'danger'
}
