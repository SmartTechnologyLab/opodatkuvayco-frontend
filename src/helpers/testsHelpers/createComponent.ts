import type { MethodOptions } from 'vue'

const createStubbedMethods = (methods: MethodOptions = {}) => {
  if (!methods) {
    return {}
  }

  return Object.keys(methods).reduce(
    (acc, key) =>
      Object.assign(acc, {
        [key]: () => {}
      }),
    {}
  )
}

export const stubComponent = (Component: any, options: Record<string, unknown> = {}) => {
  return {
    props: Component.props,
    model: Component.model,
    methods: createStubbedMethods(Component.methods),
    // Do not render any slots/scoped slots except default
    // This differs from VTU behavior which renders all slots
    template: '<div><slot></slot></div>',
    // allows wrapper.find(Component) to work for stub
    $_vueTestUtils_original: Component,
    ...options
  }
}
