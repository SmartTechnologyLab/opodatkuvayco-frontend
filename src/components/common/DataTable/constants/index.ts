import { TableSize } from '@/components/TestDataTableForUser/common/constants'
import type { TableHeaders } from '../types'
import { Currency } from '@/constants/currencies'

export enum FormatType {
  Date,
  Number,
  CurrencyUAH,
  CurrencyUSD,
  CurrencyEUR,
  Percent
}

export const dynamicCurrencies = [FormatType.CurrencyUSD, FormatType.CurrencyEUR]

export const currenciesName: Record<Currency, Currency> = {
  [Currency.UAH]: Currency.UAH,
  [Currency.USD]: Currency.USD,
  [Currency.EUR]: Currency.EUR
}

export const resultHeaders: TableHeaders[] = [
  {
    header: 'Тікер',
    field: 'ticker',
    size: [TableSize.LG, TableSize.SM],
    sortable: true
  },
  {
    header: 'Дата покупки',
    field: 'purchase.date',
    type: FormatType.Date,
    size: [TableSize.LG]
  },
  {
    header: 'Кількість',
    field: 'quantity',
    type: FormatType.Number,
    size: [TableSize.LG]
  },
  {
    header: 'Ціна',
    field: 'purchase.price',
    type: FormatType.CurrencyUSD,
    size: [TableSize.LG]
  },
  {
    header: 'Сума',
    field: 'purchase.sum',
    type: FormatType.CurrencyUSD,
    size: [TableSize.LG]
  },
  {
    header: 'Комісія',
    field: 'purchase.commission',
    type: FormatType.CurrencyUSD,
    size: [TableSize.LG]
  },
  {
    header: 'Курс',
    field: 'purchase.rate',
    type: FormatType.CurrencyUAH,
    size: [TableSize.LG]
  },
  {
    header: 'Покупка',
    field: 'purchase.uah',
    type: FormatType.CurrencyUAH,
    size: [TableSize.LG, TableSize.SM]
  },
  {
    header: 'Дата продажу',
    field: 'sale.date',
    type: FormatType.Date,
    size: [TableSize.LG]
  },
  {
    header: 'Ціна',
    field: 'sale.price',
    type: FormatType.CurrencyUSD,
    size: [TableSize.LG]
  },
  {
    header: 'Сума',
    field: 'sale.sum',
    type: FormatType.CurrencyUSD,
    size: [TableSize.LG]
  },
  {
    header: 'Комісія',
    field: 'sale.commission',
    type: FormatType.CurrencyUSD,
    size: [TableSize.LG]
  },
  {
    header: 'Курс',
    field: 'sale.rate',
    type: FormatType.CurrencyUAH,
    size: [TableSize.LG]
  },
  {
    header: 'Продаж',
    field: 'sale.uah',
    type: FormatType.CurrencyUAH,
    size: [TableSize.LG, TableSize.SM]
  },
  {
    header: 'Всього',
    field: 'total',
    type: FormatType.CurrencyUAH,
    size: [TableSize.LG, TableSize.SM]
  },
  {
    header: '%',
    field: 'percent',
    type: FormatType.Percent,
    size: [TableSize.LG]
  }
]
