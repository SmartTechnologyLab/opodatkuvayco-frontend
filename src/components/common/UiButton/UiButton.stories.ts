import UiButton from './UiButton.vue'
import { Icons, Severities } from './constants'
import type { Meta, StoryFn } from '@storybook/vue3'

export default {
  component: UiButton,
  args: {
    hasSlot: false,
    label: 'Button',
    loading: false,
    disabled: false
  },
  argTypes: {
    onClick: { action: 'click' },
    icon: {
      control: 'radio',
      options: [undefined, ...Object.values(Icons)],
      description: 'Choose icon of the button'
    },
    severity: {
      control: 'radio',
      options: Object.values(Severities),
      description: 'Choose icon of the button'
    }
  }
} as Meta<typeof UiButton>

const Template: StoryFn<typeof UiButton> = (args) => ({
  components: { UiButton },
  setup() {
    return { args }
  },
  template: `<div>
    
    <UiButton v-if="args.hasSlot" v-bind="args">{{ $t(args.label) }}</UiButton>
    <UiButton v-else  v-bind="args" />
  </div>
  `
})

export const Default = Template.bind({})
Default.args = {
  icon: undefined,
  severity: Severities.SECONDARY,
  rounded: true,
  outlined: true,
  raised: true
}

export const Colors: StoryFn<typeof UiButton> = (args) => ({
  components: { UiButton },
  setup() {
    return { args, Severities }
  },
  template: `<div style="display: grid; gap: 8px; grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));">
    <UiButton v-for="severity in Severities" :key="severity" v-bind="{ ...args, severity, label: severity}" />
  </div>`
})

export const Ico: StoryFn<typeof UiButton> = (args) => ({
  components: { UiButton },
  setup() {
    return { args, Icons }
  },
  template: `<div style="display: flex; gap: 8px;">
    <UiButton v-for="icon in Icons" :key="icon" v-bind="{ ...args, icon, label: ''}" />
  </div>`
})
