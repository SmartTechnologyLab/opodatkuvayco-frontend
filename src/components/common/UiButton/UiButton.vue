<script setup lang="ts">
import Button from 'primevue/button'
import { computed, useAttrs } from 'vue'
import { Icons, Severities } from './constants'

defineOptions({
  inheritAttrs: false
})

const attrs = useAttrs()

const props = defineProps<{
  icon?: Icons
  severity?: Severities
  rounded?: boolean
  outlined?: boolean
  raised?: boolean
}>()

defineEmits<{
  (e: 'click'): void
}>()

const routerLinkTag = 'router-link'
const linkTag = 'a'

const component = computed(() => {
  if (attrs.to) {
    return routerLinkTag
  }
  if (attrs.href) {
    return linkTag
  }

  return null
})

const componentAttrs = computed(() => {
  const { href, to } = attrs

  if (component.value === 'router-link') {
    return { to }
  }

  return { href }
})
</script>

<template>
  <Component :is="component" v-if="component" v-bind="{ ...componentAttrs }">
    <Button v-bind="{ ...props, ...attrs }" @click="$emit('click')">
      <slot />
    </Button>
  </Component>

  <Button v-bind="{ ...props, ...attrs }" @click="$emit('click')" v-else>
    <slot />
  </Button>
</template>
