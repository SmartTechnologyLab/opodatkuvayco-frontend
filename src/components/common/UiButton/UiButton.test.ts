import { RouterLinkStub, VueWrapper, shallowMount } from '@vue/test-utils'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import UiButton from '@/components/common/UiButton/UiButton.vue'
import { createRouterMock, injectRouterMock } from 'vue-router-mock'
import { Icons, Severities } from './constants'
import { stubComponent } from '@/helpers/testHelpers/createComponent'

const router = createRouterMock({
  spy: {
    create: vi.fn,
    reset: vi.fn
  }
})

type Props = {
  icon?: Icons
  severity?: Severities
  rounded?: boolean
  outlined?: boolean
  raised?: boolean
}

const ButtonStub = stubComponent(UiButton, {
  template: '<button><slot /></button>'
})

describe('UiButton component', () => {
  let wrapper: VueWrapper<InstanceType<typeof UiButton>>

  const createComponent = (props?: Props | null, attrs?: Record<string, unknown>) => {
    wrapper = shallowMount(UiButton, {
      props,
      attrs,
      slots: {
        default: 'Submit'
      },
      global: {
        stubs: {
          Button: ButtonStub,
          RouterLink: RouterLinkStub
        }
      }
    })
  }

  beforeEach(() => {
    injectRouterMock(router)
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('renders without props', () => {
    createComponent()
    expect(wrapper.exists()).toMatchSnapshot()
  })

  it('renders with provided props', async () => {
    const CHECK_ICON = 'pi pi-check'
    const SEVERITY_DANGER = 'danger'

    const buttonProps = {
      icon: CHECK_ICON,
      severity: SEVERITY_DANGER,
      rounded: true,
      outlined: true,
      raised: true
    } as Props

    createComponent(buttonProps)

    expect(wrapper.props().icon).toMatchSnapshot()
    expect(wrapper.props().severity).toMatchSnapshot()
    expect(wrapper.props().rounded).toMatchSnapshot()
    expect(wrapper.props().outlined).toMatchSnapshot()
    expect(wrapper.props().raised).toMatchSnapshot()
  })

  it('emits click event when button get clicked', async () => {
    createComponent()

    await wrapper.trigger('click')

    expect(wrapper.emitted()).toMatchSnapshot()
  })

  it('passes value to defaul slot', () => {
    createComponent()

    expect(wrapper.text()).toMatchSnapshot()
  })

  const TEST_ROUTE = '/home'
  it('have parent element of router link if it has to attribute', async () => {
    createComponent(null, {
      to: '/'
    })

    await wrapper.trigger('click')

    await wrapper.router.push(TEST_ROUTE)

    expect(wrapper.findComponent(RouterLinkStub).exists()).toMatchSnapshot()
    expect(wrapper.router.currentRoute.value.path).toMatchSnapshot()
  })

  it('have not attribute class on element router link', async () => {
    createComponent()
    const CLASS_BTN = 'submit-btn'

    wrapper = shallowMount(UiButton, {
      attrs: {
        to: TEST_ROUTE,
        class: CLASS_BTN
      }
    })

    expect(wrapper.findComponent({ name: 'RouterLink' }).attributes().class).toMatchSnapshot()
  })

  const HREF_TEST = '#about'
  it('have parent element of anchor if it has href attribute', async () => {
    createComponent()
    wrapper = shallowMount(UiButton, {
      attrs: {
        href: HREF_TEST
      }
    })

    expect(wrapper.find('a').attributes().href).toMatchSnapshot()
  })

  it('have not attribute class on element a', async () => {
    createComponent()
    const CLASS_BTN = 'submit-btn'

    wrapper = shallowMount(UiButton, {
      attrs: {
        href: HREF_TEST,
        class: CLASS_BTN
      }
    })

    expect(wrapper.find('a').attributes().class).toMatchSnapshot()
    expect(wrapper.findComponent({ name: 'Button' }).attributes().class).toMatchSnapshot()
  })
})
