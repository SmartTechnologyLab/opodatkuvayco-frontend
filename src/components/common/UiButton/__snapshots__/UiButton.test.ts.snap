// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UiButton component > emits click event when button get clicked 1`] = `
{
  "click": [
    [],
  ],
}
`;

exports[`UiButton component > have not attribute class on element a 1`] = `undefined`;

exports[`UiButton component > have not attribute class on element a 2`] = `"submit-btn"`;

exports[`UiButton component > have not attribute class on element router link 1`] = `undefined`;

exports[`UiButton component > have parent element of anchor if it has href attribute 1`] = `"#about"`;

exports[`UiButton component > have parent element of router link if it has to attribute 1`] = `true`;

exports[`UiButton component > have parent element of router link if it has to attribute 2`] = `"/home"`;

exports[`UiButton component > passes value to defaul slot 1`] = `"Submit"`;

exports[`UiButton component > renders with provided props 1`] = `"pi pi-check"`;

exports[`UiButton component > renders with provided props 2`] = `"danger"`;

exports[`UiButton component > renders with provided props 3`] = `true`;

exports[`UiButton component > renders with provided props 4`] = `true`;

exports[`UiButton component > renders with provided props 5`] = `true`;

exports[`UiButton component > renders without props 1`] = `true`;
